rules/rules.libsonnet

```diff
@@ -148,6 +148,37 @@
             },
           }
           for quantile in ['0.99', '0.9', '0.5']
+        ] + [
+          {
+            record: 'code:apiserver_request_count:rate:sum',
+            expr: |||
+              sum(
+                rate(
+                  apiserver_request_count{%(kubeApiserverSelector)s}[5m]
+                )
+              ) by (code)
+            ||| % $._config,
+          },
+          {
+            record: 'code_ratio:apiserver_request_count',
+            expr: |||
+              (
+                (
+                  sum_over_time(
+                    (sum(code:apiserver_request_count:rate:sum))[%(lookbackPeriod)s:1s]
+                  ) -
+                  sum_over_time(
+                    (sum(code:apiserver_request_count:rate:sum{code=~'5.*'}))[%(lookbackPeriod)s:1s]
+                  )
+                ) /
+                sum_over_time(
+                  (sum(code:apiserver_request_count:rate:sum))[%(lookbackPeriod)s:1s]
+                )
+              ) OR (
+                absent(code:apiserver_request_count:rate:sum{code=~'5.*'} == 0)
```

Comment:
Are we sure this does what we think? My understanding is that you're trying to guard here from the case where no 5xx code exists (not even the time-series), in which case "availability" would be 100%, do I understand correctly? (if so I think this would need to be the first OR side, not the latter, and no `== 0`)

Reply:
Yes, your understanding is correct. Let me do some testing and get back to you soon on this one

Reply:
![Screenshot from 2019-07-12 10-25-26](https://user-images.githubusercontent.com/5934711/61135616-01907100-a490-11e9-96c8-df7950f4d993.png)
Tried flipping the operands to the OR, and it doesn't seem to cause any changes (see screenshot above; timespan where there's a flat-line is where the `absent()` function takes over. 

As for the `== 0`, I remember I originally did that because some extraneous labels were creeping into some other related query, but I can't seem to replicate that issue now with this query. So I'll remove that

Reply:
Right, that's when both 5xx _and_ the other series don't have results, but will the later OR take precedence when only 5xx doesn't exist or is 0 (potentially causing a NaN)? I'm not sure myself I would have to try as well. Maybe it's worth writing a test for this.

Reply:
![Screenshot from 2019-07-12 13-14-04](https://user-images.githubusercontent.com/5934711/61146015-f432b100-a4a6-11e9-934d-6897a2751900.png)

Took some digging, but I found an example of this condition where a time series for a "bad" code doesn't exist but "good" code time series do. The cluster I'm testing on always has the 5xx codes reporting a non-zero value, but it had a "0" status code (not sure why) series that only started to appear about 1 minute after the rest of the metrics. So I slightly modified the rule query to use 0 codes instead of 5xx code for the purposes of this test.

On the left is the query in question (with `absent()` after the `OR`), on the right is `code:apiserver_request_count:rate:sum`, but with only the `code=200` and `code=0` series showing. Both graphs are looking at the same timeframe. If we look at the vertical labeled `50`, we see that only the `{code=200}` series exists on the right graph, and that the left graph is (correctly) showing the query equals 100% because the `absent()` operand (green line) dominates. 

Now looking at the vertical labeled `51`, we see the `{code=0}` series (orange line) has come into existence on the right graph, and that the left graph is (again, correctly) showing the query result as less than 100% because the `sum_over_time()` operand (orange line) dominates.

On a separate note, this test helped me to realize why we need the `==0` inside the `absent()` function. Without it, the `code` label sneaks into the result whenever the `absent()` operand dominates, which is why the left graph is showing two separate series instead of one smooth line.